5-WebApp-AuthZ-Roles-Groups/README.md

```diff
@@ -0,0 +1,252 @@
+
+### Step 3: Run the sample
+
+1. Clean the solution, rebuild the solution, and run it.
+
+2. Open your web browser and make a request to the app. The app immediately attempts to authenticate you via the Microsoft identity platform (formerly Azure AD v2.0) endpoint. Sign in with a work or school account.
+
+1. On the home page, the app lists the various claims it obtained from your ID token. You'd notice one more claims named `groups`. If **Overage** occurred, then you'd see a different claim by the name `_claim_names`. The **Overage** scenario is discussed in detail below. 
+1. On the top menu, click on the signed-in user's name **user@domain.com**, you should now see all kind of information about yourself including your picture. Beneath that, a list of all the security groups that the signed-in user is assigned to are listed as well. All of this was obtained by making calls to Microsoft Graph. This list is useful, if an **Overage** occurs with this signed-in user. The **overage** scenario is discussed later in this article. 
+
+> Did the sample not work for you as expected? Did you encounter issues trying this sample? Then please reach out to us using the [GitHub Issues](../../../../issues) page.
+
+#### Processing Groups claim in tokens, including handling **overage**.
+
+#### The `groups` claim.
+
+The object id of the security groups the signed in user is member of is returned in the `groups` claim of the token.
+
+```JSON
+{
+  ...
+  "groups": [
+    "0bbe91cc-b69e-414d-85a6-a043d6752215",
+    "48931dac-3736-45e7-83e8-015e6dfd6f7c",]
+  ...
+}
+```
+
+#### Groups overage claim
+
+To ensure that the token size doesn’t exceed HTTP header size limits, the Microsoft Identity Platform limits the number of object Ids that it includes in the **groups** claim.
+
+If a user is member of more groups than the overage limit (**150 for SAML tokens, 200 for JWT tokens**), then the Microsoft Identity Platform does not emit the groups claim in the token. Instead, it includes an **overage** claim in the token that indicates to the application to query the [Graph API](https://graph.microsoft.com) to retrieve the user’s group membership.
+
+```JSON
+{
+  ...
+  "_claim_names": {
+    "groups": "src1"
+    },
+    {
+   "_claim_sources": {
+    "src1": {
+        "endpoint":"[Graph Url to get this user's group membership from]"
+        }
+    }
+  ...
+}
+```
+> An Identity Office Hours session covered Azure AD App roles and security groups, featuring this scenario and how to handle the overage claim. Watch the video [Using Security Groups and Application Roles in your apps](https://www.youtube.com/watch?v=V8VUPixLSiM)
+
+ > You can use the `BulkCreateGroups.ps1` provided in the [App Creation Scripts](./AppCreationScripts/) folder to help test overage scenarios.
+
+##### Order of processing the overage claim
+
+1. Check for the claim `_claim_names` with one of the values being `groups`. This indicates overage.
```

Comment:
Academically yes, but the code, for example, in C#, has to be written as following..

            // first check for _claim_names claim, when this present, the 'groups' claim will not be present "at root", it will be a sub-claim under _claim_names.
            if (claimsPrincipal.FindFirstValue("_claim_names") == null)
            {
                // Process groups
                var groups = claimsPrincipal.FindFirstValue("groups");
            }
            else
            {
                // Check for 'groups' claim being present in _Claim_names. Find the groups 'src' and make a call to the source provided in _claim_sources
            }

---
In reply to: [273175741](https://github.com/Azure-Samples/active-directory-aspnetcore-webapp-openidconnect-v2/pull/84#discussion_r273175741) [](ancestors = 273175741)